### DISCUSSION TITLE
Is there a way to redefine \int to \int\limits globally?

### DISCUSSION THREAD

**Discussion by pedrolimasi** [KaTeX/KaTeX#3736] [Q&A] [answered]
Setting global macros is easy enough, but macros requires 2 things for a substitution. What can we do for something done with `\let`?

As a simple example:
```latex
\let\Oldint=\int
\def\int{\Oldint\limits}
```

This works isolated equation-by-equation, but I'd like to set it globally.
The second line has a macro syntax, but not the first.

There's an open issue (#2915)  that is somewhat related but not quite the same.

**Comment by edemaine** [comment]
You could run the macro code with `globalGroup: true` and pass in a `macros` object, then provide that `macros` object to all future calls. If you're not familiar with the API (e.g. you're just using auto-render), I can provide more explicit code.

**Reply by pedrolimasi** [reply]
I'm running it through [a VSCode extension](https://github.com/shd101wyy/mume "mume + Markdown Preview Enhanced"). They expose enough of the interface to KaTeX for me to define the `macros` object.

From a quick glance at their code, it seems they're passing the `macros` object as you suggest:
https://github.com/shd101wyy/mume/blob/7d2d7956bbd3e31ebaecb6efe20200cadbd7fdc2/src/parse-math.ts#L47-L50

However, when I try to use `\let` to define a macro there, it doesn't work.

> If you're not familiar with the API (e.g. you're just using auto-render), I can provide more explicit code.

More explicit code may help. Thanks!

**Reply by pedrolimasi** [reply]
Perhaps a sub-question may be the key here:

How should I add `\Oldint` as a macro in the `macros` object?
- `"\\let\\Oldint": "\\int"` doesn't work, and
- `"\\Oldint": "\\int"` creates an infinite loop when I redefine `\int`

**Reply by edemaine** [reply]
~~Apologies; I thought this was possible, but it turns out it's not. I just tried ``katex.renderToString(`\let\Oldint=\int`, {macros, globalGroup: true})`` with the goal of looking at what `macros` object gets produced, but it turns out that this doesn't populate the `macros` object at all.~~

**Reply by edemaine** [reply]
Sorry, I made a mistake. Instead I did this:
```js
katex.renderToString("\\let\\Oldint=\\int", {macros, globalGroup: true}) 
```
Now `macros` gets populated as roughly `{"\\Oldint": {tokens: [{text: "\\int", noexpand: true}], unexpandable: false, numArgs: 0}}`.

I would therefore expect `macros: {"\\Oldint": {tokens: [{text: "\\int", noexpand: true}], unexpandable: false, numArgs: 0}, "\\int": "\\Oldint\\limits"}` to work... but currently it seems to produce an infinite loop.  Hmm.

**Reply by pedrolimasi** [reply]
Is this enough to open an issue? Since I'm running KaTeX through the VSCode extension [Markdown Preview Enhanced](https://github.com/shd101wyy/markdown-preview-enhanced), I'd expect a bug report from my use-case description would be... bad.

**Reply by edemaine** [reply]
Yep. Just opened #3737. Thanks!

**Comment by pedrolimasi** [comment, marked as answer]
[WORKAROUND]

Use `\intop` instead of `\int\limits`.

In the macros object, that would be:
`"\\int": "\\intop"`
